src/java/org/jivesoftware/openfire/archive/ArchiveInterceptor.java

```diff
@@ -89,11 +90,42 @@ public void interceptPacket(Packet packet, Session session, boolean incoming, bo
                         JID sender = message.getFrom();
                         JID receiver = message.getTo();
                         ConversationEventsQueue eventsQueue = conversationManager.getConversationEventsQueue();
-                        eventsQueue.addChatEvent(conversationManager.getConversationKey(sender, receiver),
+                        if (message.getBody()!=null)
+                        {
+                            eventsQueue.addChatEvent(conversationManager.getConversationKey(sender, receiver),
                                 ConversationEvent.chatMessageReceived(sender, receiver,
                                         conversationManager.isMessageArchivingEnabled() ? message.getBody() : null,
                                         conversationManager.isMessageArchivingEnabled() ? message.toXML() : null,
                                         new Date()));
+                        }
+                        else
+                        {
+                            EmptyMessageType emptyMessageType = EmptyMessageUtils.getMessageType(message.getElement()); 
+
+                            long bitmask = conversationManager.getSpeficifEmptyMessageArchivingEnabled();
+
+                            if (emptyMessageType!=EmptyMessageType.TYPE_EVENT && (
+                                emptyMessageType==EmptyMessageType.TYPE_UNKNOWN && ((bitmask & EmptyMessageType.TYPE_UNKNOWN.getValue())==EmptyMessageType.TYPE_UNKNOWN.getValue())||
```

Comment:
This bitmasking implementation is overly complex (it is hard to reason over). I don't think it is doing what it intends to do: the emptyMessageType of the stanza is compared with just one bitmask value.

Reply:
I've fixed and refactored this.